Add Baseline definition document (#423)
web-platform-dx/web-features: docs/baseline.md

_Baseline_ (alongside related statuses) is intended to help these developers by offering a shortcut past the complexity.

## Goals

_Baseline_ status is a wayfinding tool for web developers.
Baseline status aims to help web developers to identify consensus web platform features and distinguish them from features that are in other states of development (e.g., proposals, experiments, or deprecated and historic features; see also [Future considerations](#future-considerations)).
In other words, Baseline status aims to identify features of developers’ contemporary, workaday web.

Goal 1: **Tailor Baseline status to a developer audience.**
The Baseline status is intended for developers creating and maintaining hypertext applications for the web, who may not have access to the resources, data, tools, or knowledge needed to identify their users’ browsers capabilities more directly.
See [Audience illustration for Baseline status](#audience-illustration-for-baseline-status) for a detailed story to help understand this goal in more detail.

> Goal 2: **Change with the web and web developers’ needs.** Features can and should join and leave Baseline status as their respective interoperability, reliability, commitment, and availability profiles change. Likewise, Baseline statuses should change as developers’ needs change (for example, developers come to rely on a new browser or an established browser fades from use). Baseline is not an inert, one-time assessment.

Author: ```suggestion
Goal 2: **Change with the web and web developers’ needs.**
Features can and should join and leave Baseline status as their respective interoperability, reliability, commitment, and availability profiles change.
Likewise, Baseline statuses should change as developers’ needs change (for example, developers come to rely on a new browser or an established browser fades from use).
Baseline is not an inert, one-time assessment, nor is it arbitrary, changing form from one day to the next.
```

Something like this, maybe? (last line)

> Goal 2: **Change with the web and web developers’ needs.** Features can and should join and leave Baseline status as their respective interoperability, reliability, commitment, and availability profiles change. Likewise, Baseline statuses should change as developers’ needs change (for example, developers come to rely on a new browser or an established browser fades from use). Baseline is not an inert, one-time assessment.

Reviewer: Neat! 👍

Baseline features demonstrate consensus through constancy: uninterrupted support from one release to the next.
Baseline is not for prototypes, betas, or temporary hacks.
All software has bugs, but Baseline features have narrow, known bugs.

Goal 6: **Identify commitment through documentation.**
Baseline features demonstrate consensus through specifications, documentation for web developers, and other signs that the feature will continue to have a place in a web developer’s toolbox.
Baseline features aren’t unspecified or formally discouraged.

<!-- TODO: https://github.com/web-platform-dx/web-features/issues/437 -->

## Non-goals

Baseline status cannot or will not satisfy the following non-goals:

* **Identify universally available features.**
  Many Baseline features will not achieve 100% user reach soon or perhaps ever.
  If a web developer needs to support a globally uncommon or discontinued browser (e.g., Internet Explorer 11), then the developer needs to know the specific limitations of that browser, not a broad overview of developers’ most commonly required browsers.
  Baseline can’t be both.
* **Identify support in non-web environments.**

>   Developers use web technologies in non-web settings, such as JavaScript in Node.js, Web APIs in Deno, or HTML and CSS in Electron-based applications.

Author: Ah, this is a good thought though I don't know exactly where it sits either. I think it's more of a future consideration than a non-goal, but it's possible it could _become_ a non-goal. I filed https://github.com/web-platform-dx/web-features/issues/477 to take up this question.

   - The governance group is withholding or modifying the status (i.e., there is no editorial override of the feature’s status).

If the feature has the interoperable status, then the feature’s _keystone date_ shall be set to the last release date on which a browser introduced support for the feature
If there was more than one introduction (e.g., a feature was withdrawn then reintroduced), then only the latest date applies.

### Wider-support (high) status

For each feature definition in `web-features`, a wider-support status shall be be conferred to the feature if the feature satisfies the following test:

The feature’s keystone date is on or before today’s date minus 30 months and the dates of the following long-term support releases:

* Mozilla Firefox ESR, given by the release date for the latest x.0 release of Firefox ESR (or the previous x.0 release, when there are two active ESR releases).

This duration is selected to approximate developer signals, estimates of browser release uptake over time, an estimate of high total market share support, and the project governance group’s best judgment.

> This duration is due for review by the governance group on 7 November 2024.

Author: Yes, I think this is probably worth moving or revising. That said, this was one of the late additions to the Baseline definition text. I'm not comfortable approving this unilaterally. I'd like to merge the text as is, then propose the narrower change here as a separate PR for the owners group to consider in isolation.

Goal 3: **Identify interoperability through support across browser implementations.**
Baseline features demonstrate consensus through multiple implementations.
Baseline is not for single-implementation browser features.

Goal 4: **Identify availability through popular browser support.**
Baseline features demonstrate consensus through user reach, a supermajority of global users by reasonable measures used by web developers.
Baseline features should not surprise a web developer with compatibility failures.

Goal 5: **Identify continuity through support over time.**
Baseline features demonstrate consensus through constancy: uninterrupted support from one release to the next.
Baseline is not for prototypes, betas, or temporary hacks.
All software has bugs, but Baseline features have narrow, known bugs.

Goal 6: **Identify commitment through documentation.**
Baseline features demonstrate consensus through specifications, documentation for web developers, and other signs that the feature will continue to have a place in a web developer’s toolbox.
Baseline features aren’t unspecified or formally discouraged.

> <!-- TODO replace with an issue -->
> 
> **Issue TBD**: Part of goal 6 ought to be part of feature-set/web-features feature definition rules, not in Baseline status selection.

Reviewer 2: ```suggestion
<!-- TODO: https://github.com/web-platform-dx/web-features/issues/437 -->
```

  Instead, developers should be able to use Baseline to decide when to stop shipping a polyfill.
  See also: [Future considerations](#future-considerations).
* **Replace fine-grained or site-specific reporting or analyses.**
  Baseline is a complement to, not a replacement for, detailed compatibility reporting, such as _Can I Use…?_ or browser compatibility tables on MDN.
  Likewise, Baseline is not a substitute for custom analysis (e.g., [Import statistics on Can I Use…?](https://caniuse.com/ciu/import)).
  Baseline may aid web developers in directing their browser support investigations but, as a summary, won’t eliminate application-specific analysis.
* **Foreclose on other statuses.**
  Baseline’s goals capture a certain kind of stability in web platform features, but that’s far from the only status relevant to web developers.
  Other categories could surely exist to complement or work alongside Baseline.
  Baseline should seek to summarize underlying facts about web platform features, not canonize them.
  See also: [Future considerations](#future-considerations).

## Audience illustration for Baseline status

> <!-- TODO replace with an issue -->
> 
> **Issue TBD**: This persona will likely have utility for all of feature-set/web-features and should move to some project-wide documentation and get a cross reference here.

Reviewer 2: ```suggestion
<!-- TODO: https://github.com/web-platform-dx/web-features/issues/438 -->
```

> * The developer has limited time and budget to get relatively old or new devices and test with.
>   They use a 2-year old laptop that they keep _mostly_ up-to-date with OS and browser updates.
> * The developer has limited time and interest to keep up with browser news.
>   They don’t routinely read web development blogs.
> * The developer recently decided to stop worrying about end-of-life browsers.
>
> Today, to decide whether to use a new-to-them web platform feature, the developer uses the same techniques taught by their mentors: skim _Can I use…?_ and MDN browser compatibility tables.
> They mostly work on a gut feeling: is there _enough_ green in the table to use this feature?
> This works some of the time, but they’ve been occasionally surprised by both “new” (and unfamiliar) features being long-supported and unexpected complaints of incompatibility.

## Ownership and maintenance

The WebDX community group, through the [web-platform-dx/web-features-set owners group](../GOVERNANCE.md), maintains this document.

> Based on WebDX community group research, the feature-set owners group decides matters such as the core browser set, releases, editorial overrides, and so on.

Reviewer 2: ```suggestion
Based on WebDX community group research, the web-features owners group decides matters such as the core browser set, releases, editorial overrides, and so on.
```

If there was more than one introduction (e.g., a feature was withdrawn then reintroduced), then only the latest date applies.

### Wider-support (high) status

For each feature definition in `web-features`, a wider-support status shall be be conferred to the feature if the feature satisfies the following test:

The feature’s keystone date is on or before today’s date minus 30 months and the dates of the following long-term support releases:

* Mozilla Firefox ESR, given by the release date for the latest x.0 release of Firefox ESR (or the previous x.0 release, when there are two active ESR releases).

This duration is selected to approximate developer signals, estimates of browser release uptake over time, an estimate of high total market share support, and the project governance group’s best judgment.

This duration is due for review by the governance group on 7 November 2024.

### Core browser set

Both the interoperable and wider-support substatuses observe support with respect to a set of browsers.
The _core browser set_ shall be defined as:

> * Apple Safari (iOS)
> * Apple Safari (macOS)
> * Mozilla Firefox (Android)
> * Mozilla Firefox (desktop)
> * Google Chrome (Android)
> * Google Chrome (desktop)
> * Microsoft Edge (desktop)

Reviewer 2: Alphabetical:

```suggestion
* Apple Safari (iOS)
* Apple Safari (macOS)
* Google Chrome (Android)
* Google Chrome (desktop)
* Microsoft Edge (desktop)
* Mozilla Firefox (Android)
* Mozilla Firefox (desktop)
```

## Future considerations

There are many things Baseline does not cover.
Here are some areas for future consideration and not currently in-scope for Baseline, particularly other candidate states such as:

> <!-- TODO: Replace these with issues -->

Reviewer 2: Not filing issues for these, leaving them here seems fine to me.
